### PR TITLE
Modifying syscheck static fields into dynamic fields

### PR DESCRIPTION
**Pull request by jotacarma90**
|Related issue|
|---|
|https://github.com/wazuh/wazuh/issues/5002|

## Description
This PR is intended to remove all syscheck's static fields in the Eventinfo structure. Some of them were duplicated and it is simply necessary to remove the static ones, while for others it is also necessary to create the dynamic field taking care that the functionality remains the same.

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
- [x] Source installation
- [x] Review logs syntax and correct language

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Valgrind (memcheck and descriptor leaks check)
- Memory tests for Windows
  - [x] Scan-build report

### CODE REVIEW

**Comment by jotacarma90** [comment]
FIM Jenkins integration test results:
[agent_windows_html_report_test_integration_B1260_20210212150330.zip](https://github.com/wazuh/wazuh/files/5980190/agent_windows_html_report_test_integration_B1260_20210212150330.zip)
[agent_ubuntu_html_report_test_integration_B1260_20210212150330.zip](https://github.com/wazuh/wazuh/files/5980191/agent_ubuntu_html_report_test_integration_B1260_20210212150330.zip)
[agent_centos_html_report_test_integration_B1260_20210212150330.zip](https://github.com/wazuh/wazuh/files/5980192/agent_centos_html_report_test_integration_B1260_20210212150330.zip)
[agent_solaris_html_report_test_integration_B1261_20210212150335.zip](https://github.com/wazuh/wazuh/files/5980193/agent_solaris_html_report_test_integration_B1261_20210212150335.zip)
[agent_macos_html_report_test_integration_B1261_20210212150335.zip](https://github.com/wazuh/wazuh/files/5980194/agent_macos_html_report_test_integration_B1261_20210212150335.zip)

File: src/analysisd/format/to_json.c [hunk 1]
```diff
@@ -273,92 +276,91 @@ char* Eventinfo_to_jsonstr(const Eventinfo* lf, bool force_full_log)
...
         if(Config.decoder_order_size > FIM_DIFF && lf->fields[FIM_DIFF].value && strcmp(lf->fields[FIM_DIFF].value, "0")) {
             cJSON_AddStringToObject(file_diff, "diff", lf->fields[FIM_DIFF].value);
-        } else if(lf->diff && strcmp(lf->diff, "")) {
-            cJSON_AddStringToObject(file_diff, "diff", lf->diff);
         }
-        if(lf->sk_tag) {
-            if (strcmp(lf->sk_tag, "") != 0) {
+
+        if(lf->fields[FIM_TAG].value) {
+            if (strcmp(lf->fields[FIM_TAG].value, "") != 0) {
```
Comment (Molter73) [inline src/analysisd/format/to_json.c:358]:
```suggestion
        if(lf->fields[FIM_TAG].value && *lf->fields[FIM_TAG].value != '\0') {
```

File: src/unit_tests/analysisd/test_log.c [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    lf->hour[1] = '\0';
+    lf->generated_rule->sigid = 554;
+    lf->generated_rule->level = 123;
+    lf->generated_rule->group = "rule_group";
+    lf->time.tv_sec = 160987966;
+    lf->full_log = "full_log";
+    lf->comment = "comment";
+    lf->hostname = "hostname";
+
+    char aux1[12];
```
Comment (Molter73) [inline src/unit_tests/analysisd/test_log.c:111]:
```suggestion
```

File: src/unit_tests/analysisd/test_log.c [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+    for (int i = 0; i < lf->nfields; i++) {
+        lf->fields[i].value = df[i].value;
+        lf->fields[i].key = df[i].key;
+    }
```
Comment (Molter73) [inline src/unit_tests/analysisd/test_log.c:116]: Could we get away with this?
```suggestion
    lf->fields= df;
```

**Review by Molter73 — CHANGES_REQUESTED** [review]
Apart from the comments, src/analysisd/analysisd.c only has whitespace changes. In order to prevent potential conflicts with other PRs we should revert it back to how it was.

File: src/analysisd/alerts/log.c [hunk 4]
```diff
@@ -397,97 +183,91 @@ void OS_Log(Eventinfo *lf)
...
+
+        if (lf->fields[FIM_INODE].value && *lf->fields[FIM_INODE].value != '\0') {
+            fprintf(_aflog, " - Inode: %s\n", lf->fields[FIM_INODE].value);
         }
-        if (lf->fields[FIM_UID].value && lf->fields[FIM_UNAME].value) {
-            if (*lf->fields[FIM_UNAME].value) {
-                fprintf(_aflog, " - User: %s (%s)\n", lf->fields[FIM_UNAME].value, lf->fields[FIM_UID].value);
-            }
+
+        if (lf->fields[FIM_UID].value && lf->fields[FIM_UNAME].value && *lf->fields[FIM_UNAME].value != '\0') {
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply src/analysisd/alerts/log.c:207]: Just in the if below was doing that check, I unified the conditionals

File: src/analysisd/alerts/log.c [hunk 5]
```diff
@@ -397,97 +183,91 @@ void OS_Log(Eventinfo *lf)
...
+            fprintf(_aflog, " - User: %s (%s)\n", lf->fields[FIM_UNAME].value, lf->fields[FIM_UID].value);
         }
-        if (lf->fields[FIM_GID].value && lf->fields[FIM_GNAME].value) {
-            if (*lf->fields[FIM_GNAME].value) {
-                fprintf(_aflog, " - Group: %s (%s)\n", lf->fields[FIM_GNAME].value, lf->fields[FIM_GID].value);
-            }
+
+        if (lf->fields[FIM_GID].value && lf->fields[FIM_GNAME].value && *lf->fields[FIM_GNAME].value != '\0') {
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply src/analysisd/alerts/log.c:211]: Just in the if below was doing that check, I unified the conditionals

File: src/headers/syscheck_op.h [hunk 6]
```diff
@@ -117,6 +129,7 @@ typedef enum fim_fields {
     FIM_REGISTRY_VALUE_NAME,
     FIM_REGISTRY_VALUE_TYPE,
     FIM_ENTRY_TYPE,
+    FIM_EVENT_TYPE_STR,
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply src/headers/syscheck_op.h:132]: Changed the other static buffer to FIM_EVENT_TYPE_ARRAY

File: src/analysisd/decoders/syscheck.c [hunk 7]
```diff
@@ -450,29 +459,29 @@ int fim_db_search(char *f_name, char *c_sum, char *w_sum, Eventinfo *lf, _sdb *s
...
                         merror("Unable to add scan_info to hash table for agent: %s", lf->agent_id);
                     }
                 }
             } else {
                 end_scan = *end_first_scan;
             }
 
-            if(lf->event_type == FIM_ADDED) {
-                if(end_scan == 0) {
+            if (strcmp(lf->fields[FIM_EVENT_TYPE_STR].value, SYSCHECK_EVENT_STRINGS[FIM_ADDED]) == 0) {
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply src/analysisd/decoders/syscheck.c:477]: Created `syscheck_event_t event_type` and added a parameter to `fim_alert` and `fim_generate_alert` to be more efficient deleting some strcmp.

**Comment by jotacarma90** [comment]
Integration test for analysisd passed:
[manager_html_report_test_integration_B1362_20210302092909.zip](https://github.com/wazuh/wazuh/files/6074347/manager_html_report_test_integration_B1362_20210302092909.zip)
[agent_windows_html_report_test_integration_B1372_20210302144938.zip](https://github.com/wazuh/wazuh/files/6074348/agent_windows_html_report_test_integration_B1372_20210302144938.zip)
[agent_ubuntu_html_report_test_integration_B1363_20210302092912.zip](https://github.com/wazuh/wazuh/files/6074349/agent_ubuntu_html_report_test_integration_B1363_20210302092912.zip)
[agent_centos_html_report_test_integration_B1363_20210302092912.zip](https://github.com/wazuh/wazuh/files/6074350/agent_centos_html_report_test_integration_B1363_20210302092912.zip)

**Comment by jotacarma90** [comment]
Integration test in local for test_analysisd
[analysisd-integrados-local-5002.zip](https://github.com/wazuh/wazuh/files/6109859/analysisd-integrados-local-5002.zip)

File: src/analysisd/decoders/syscheck.c [hunk 8]
```diff
@@ -531,199 +544,193 @@ int fim_db_search(char *f_name, char *c_sum, char *w_sum, Eventinfo *lf, _sdb *s
...
+            changes = 1;
+            wm_strcat(&lf->fields[FIM_CHFIELDS].value, "sha1", ',');
+            snprintf(localsdb->sha1, OS_FLSIZE, "Old sha1sum was: '%s'\nNew sha1sum is : '%s'\n",
+                        oldsum->sha1, newsum->sha1);
+            os_strdup(oldsum->sha1, lf->fields[FIM_SHA1_BEFORE].value);
+        }
+
+        /* SHA-256 message */
+        if(newsum->sha256 && newsum->sha256[0] != '\0')
+        {
```
Comment (antoniomanuelfr) [inline src/analysisd/decoders/syscheck.c:675]:
```suggestion
        if(newsum->sha256 && newsum->sha256[0] != '\0')  {
```

File: src/analysisd/decoders/syscheck.c [hunk 9]
```diff
@@ -1685,44 +1679,45 @@ int fim_fetch_attributes_state(cJSON *attr, Eventinfo *lf, char new_state) {
...
-                    lf->mtime_after = (long)attr_it->valuedouble;
                 } else {
-                    lf->mtime_before = (long) attr_it->valuedouble;
+                    lf->fields[FIM_MTIME_BEFORE].value = w_long_str((long) attr_it->valuedouble);
                 }
             }
         } else if (attr_it->type == cJSON_String) {
             char **dst_data = NULL;
 
             if (!strcmp(attr_it->string, "perm")) {
```
Comment (antoniomanuelfr) [inline src/analysisd/decoders/syscheck.c:1703]: Can you change these `strcmp` to match the same structure? 
```suggestion
            if (strcmp(attr_it->string, "perm") == 0) {
```

File: src/analysisd/eventinfo.c [hunk 10]
```diff
@@ -1384,3 +1164,14 @@ void w_free_event_info(Eventinfo *lf) {
         Free_Eventinfo(lf);
     }
 }
+
+char *extract_module_from_location(const char *location){
+    char *module_name = strstr(location, "->");
```
Comment (antoniomanuelfr) [inline src/analysisd/eventinfo.c:1169]: Check if location is `NULL` before calling `strstr`

File: src/analysisd/eventinfo.c [hunk 11]
```diff
@@ -1384,3 +1164,14 @@ void w_free_event_info(Eventinfo *lf) {
...
+    if (module_name == NULL) {
+        module_name = location;
```
Comment (antoniomanuelfr) [inline src/analysisd/eventinfo.c:1171]: This line produces the following warning: 
```
analysisd/eventinfo.c: In function 'extract_module_from_location':
analysisd/eventinfo.c:1171:21: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
 1171 |         module_name = location;
      |                     ^
    CC analysisd/labels-live.o
```

**Review by antoniomanuelfr — CHANGES_REQUESTED** [review]
The PR looks good, but some style changes are requested.

**Review by antoniomanuelfr — APPROVED** [review]

**Review by Molter73 — CHANGES_REQUESTED** [review]

**Review by JoseAntonioMHerrera — DISMISSED** [review]

**Comment by jotacarma90** [comment]
test_analysisd result on Jenkins:
[manager_html_report_test_integration_B1570_20210409101812.zip](https://github.com/wazuh/wazuh/files/6285885/manager_html_report_test_integration_B1570_20210409101812.zip)

**Review by antoniomanuelfr — DISMISSED** [review]

**Review by vikman90 — APPROVED** [review]

**Pull request merged**